### DISCUSSION TITLE
Building with older Android SDKs?

### DISCUSSION THREAD

**Discussion by MetacodeX7** [AvaloniaUI/Avalonia#11331] [Q&A] [answered]
Hi,

I've been trying without success to build my Avalonia app using an older Android SDK, such as API 29 and API 30. 

My project works fine with API 33 but my phone isn't compatible. I change the target platform in the Android project and the manifest, but Visual studio (2022 17.5.5) keeps telling me that the workload is missing.

`error NETSDK1181: Error getting pack version: Pack 'Microsoft.Android.Ref.30' was not present in workload manifests.`

I tried maui-check without success. This only installs the latest android SDK, not older ones.
I tried all kinds of things found on the internet or chatgpt without success. I can display the workload, but I don't see how to modify it. It has been suggested to install from the Visual Studio Installer in order to affect the workload, but the installer only offers to install API 33.

I have all 3 SDKs installed with the Android SDK Manager in my `C:\Program Files (x86)\Android\android-sdk\platforms` folder, and this folder is the one used in `Options > Xamarin > Android Settings`. But it's as if it only ever finds the API 33 sdk, fully trusting the workload specified by the VS Installer, and completely ignoring the Xamarin SDK folder.

Has anyone managed to do this? How can I make it work?

**Comment by jp2masa** [comment]
> My project works fine with API 33 but my phone isn't compatible.

IIRC, the target SDK only represents the API you target, not the minimum SDK version. So it should be possible to change the minimum SDK version to the one you need (I think it's in the app manifest?).

**Reply by MetacodeX7** [reply]
It is possible to target API 33 and run on API 30 devices anyway, and I have done so. But this doesn't work for API 29, which I think is incompatible. Even with the API 30 device, I encounter an unknown crash on startup in Release mode (no crash in Debug), so I want to build it properly by targeting API 30 to make sure this isn't related.

**Reply by jp2masa** [reply]
Did you try running `dotnet workload restore`?

**Reply by MetacodeX7** [reply]
I believe I did. It took a long while but did not install another android SDK, since nothing is telling it to modify the workload before restoring it.

**Comment by maxkatz6** [comment, marked as answer]
You need a newer Android SDK and project target version. Note, some newer SDK versions might also require you to have a newer .NET SDK version.
If you want to support older android versions, you can use SupportedOSPlatformVersion.
Avalonia technically supports 21+ https://github.com/AvaloniaUI/Avalonia/blob/master/src/Android/Avalonia.Android/Avalonia.Android.csproj#LL4C6-L4C32 (though it wasn't tested with real devices such old, real min version might be higher).

> But this doesn't work for API 29, which I think is incompatible. Even with the API 30 device, I encounter an unknown crash on startup in Release mode (no crash in Debug), so I want to build it properly by targeting API 30 to make sure this isn't related.

It is compatible. At least should be.
If you have any avalonia related crashes, please report it here. If seems like a runtime/sdk bug - report it in https://github.com/xamarin/xamarin-android repository (you also can search there for some hints).

**Reply by MetacodeX7** [reply]
> It is compatible. At least should be.

You were right. At some point during my tests the API 29 devices appeared incompatible in Visual Studio, and I found some information to explain it, but that must have been mistaken. I tried it again and this time I was able to run an API 33 build on my API 29 phone. Saves me a lot of headaches, thanks!

> If you have any avalonia related crashes, please report it here.

Unfortunately I have no clue what the crash is related to yet. There isn't much in the logs that seems relevant, even after turning on some debug options in Release.
